File: cumulative_distgen.py
############################################################################
# Cumulative Distribution Function for Arbitrary Probability Distributions #
# Glenn Dawson                                                             #
# 2017-10-06                                                               #
############################################################################

###################
# List of Imports #
###################
import random
import numpy as np
import matplotlib.pyplot as plt

########
# Main #
########
def main():
    random_discrete_distribution = np.random.rand(400,1)
    random_discrete_distribution = random_discrete_distribution / np.sum(random_discrete_distribution)
    sample = cumulative_dist_gen(random_discrete_distribution,10000)
    plt.hist(sample,100)
    plt.show()

##################################################################
# Cumulative Distribution Function                               #
# -------------                                                  #
# Arguments:                                                     #
# X is a vector containing normalized probabilities              #
# Y is an integer representing the number of samples to be drawn #
##################################################################
def cumulative_dist_gen(X,Y):
    sample = []
    for i in range (0,Y):
        r = random.uniform(0,1)
        cumdist = 0
        j=0
        while (r >= cumdist):
            cumdist = cumdist + X[j]
            j += 1
        sample.append(j)
    return sample

if __name__ == '__main__':
    main()

File: MATLAB/cumulative_distgen.m

% ******************************************************************* %
% Cumulative Distribution Function of Arbitrary Discrete Distribution %
% Glenn Dawson                                                        %
% 2017-09-19                                                          %
% ******************************************************************* %

% Arguments:
% X is a vector containing normalized probabilities
% Y is an integer representing the number of samples to be drawn from X
% Returns:
% 1-by-Y vector m ~ X

function m=cumdistgen(X, Y)

% Generate a sample from the distribution
sample = [];
for i=1:Y
    r=rand;                         % Generate a random number {0 1}
    cumdist=0;                      % Reset cumulative distribution
    j=1;                            % Counter for X
    while r >= cumdist              
        cumdist = cumdist + X(j);   % Add X(j) probability to extant sum
        j = j+1;                    % Increment to X(j+1)
    end
    sample = [sample (j-1)];        % Add chosen element of X to sample
end

% Return sample
m=sample; 